### PR TITLE
Generate hash ID to make for security policy rule ID unique

### PR DESCRIPTION
**Pull request by timdengyun**
In certain case, SecurityPolicy cannot delete a rule when this rule is not the last one.
For example, one SP with rule1, rule2 and rule3.
the rule ID is sp_spUID_0, sp_spUID_1, and sp_spUID_2, respectively.

When user is trying to delete rule 2, the deleting will fail. This issue is caused by the original rule2 ID: sp_spUID_1_0_0 will be regenerated by the original rule3, because rule3 becomes the new rule2, the new rule2 id will be sp_spUID_2. NSX will complain that groups in the deleted rule2 are still being referenced by sp_spUID_2.

The root cause for the issue is that we don't generate unique ID for the each rule.

Hence, this patch is to
1.Generate hash ID for security policy rule content and append this hash ID to the final generated rule ID to make each rule's ID unique. 
So, the final generated rule ID is: sp_spUID_rulehashID_ruleIndex.

Also, plus following refactor changes:
2. Unify the applied group ID/name generation for policy and rule level.
3. Unify the source/destination group ID/name generation for rule level.

The testing Done:
1. Creating SP with rule1, rule2 and rule 3, deleting the rule2.
2. Creating SP with rule1, rule2 and rule 3, deleting the rule1.
3. Creating SPs without this patch, and staring NSX Operator again with this patch to verify that new generated rules with unique rule ID are created successfully. This is to test upgrade existing SPs.

### CODE REVIEW

File: pkg/nsx/services/builder.go [hunk 1]
```diff
@@ -294,21 +307,46 @@ func (service *SecurityPolicyService) buildExpressionsMatchExpression(matchExpre
...
+}
+
+// build appliedTo group path for both policy and rule levels.
+func (service *SecurityPolicyService) buildAppliedGroupPath(obj *v1alpha1.SecurityPolicy, ruleIdx int) string {
+	var groupID string
+	if ruleIdx == -1 {
+		groupID = service.buildAppliedGroupID(obj, -1)
+	} else {
+		groupID = service.buildAppliedGroupID(obj, ruleIdx)
+	}
```
Comment (heypnus) [inline pkg/nsx/services/builder.go:340]: It looks that L335-L340 can be simplified to:
`groupID = service.buildAppliedGroupID(obj, ruleIdx)`

File: pkg/nsx/services/builder.go [hunk 2]
```diff
@@ -374,33 +408,24 @@ func (service *SecurityPolicyService) buildRuleAndGroups(obj *v1alpha1.SecurityP
...
+	)
+	if err != nil {
+		return nil, nil, err
 	}
+	ruleGroups = append(ruleGroups, nsxRuleAppliedGroup)
 	nsxRule.Scope = []string{nsxRuleAppliedGroupPath}
 
-	log.V(1).Info("built rule and groups", "nsxRuleAppliedGroup", nsxRuleAppliedGroup, "nsxRuleSrcGroup", nsxRuleSrcGroup, "nsxRuleDstGroup", nsxRuleDstGroup, "action", *nsxRule.Action, "direction", *nsxRule.Direction)
+	log.V(1).Info("built rule and groups", "nsxRuleAppliedGroup", nsxRuleAppliedGroup,
+		"~", nsxRuleSrcGroup, "nsxRuleDstGroup", nsxRuleDstGroup,
```
Comment (heypnus) [inline pkg/nsx/services/builder.go:425]: `"~"`?

File: pkg/nsx/services/builder.go [hunk 3]
```diff
@@ -839,12 +845,13 @@ func (service *SecurityPolicyService) mergeSelectorMatchExpression(matchExpressi
...
-	var mexprInOpCount = 0
-	var mexprInValueCount = 0
+	matchLabels map[string]string,
+) (int, error) {
+	mexprInOpCount := 0
+	mexprInValueCount := 0
 	var err error = nil
-	var errorMsg = ""
-	var exists = false
+	errorMsg := ""
```
Comment (heypnus) [inline pkg/nsx/services/builder.go:853]: Why we changed all `var errorMsg = ""` to `errorMsg := ""`?

File: pkg/nsx/services/builder.go (hunk 1 above)
Comment (timdengyun) [reply pkg/nsx/services/builder.go]: Thanks, good point. I've changed the code as advised.

File: pkg/nsx/services/builder.go (hunk 2 above)
Comment (timdengyun) [reply pkg/nsx/services/builder.go]: Thanks for bringing out the accident change in the previous patch.  I've changed the code to fix this error.

File: pkg/nsx/services/builder.go (hunk 3 above)
Comment (timdengyun) [reply pkg/nsx/services/builder.go:850]: This sort of changes were done by golangci-lint tool in vscode. I think the tool prefers this code style way.

**Review by heypnus — DISMISSED** [review]

**Review by heypnus — APPROVED** [review]

**Pull request merged**